README.md
# Puppeteer Cluster

[![Build Status](https://github.com/thomasdondorf/puppeteer-cluster/actions/workflows/actions.yml/badge.svg)](https://github.com/thomasdondorf/puppeteer-cluster/actions/workflows/actions.yml)
[![npm](https://img.shields.io/npm/v/@zhaow-de/puppeteer-cluster)](https://www.npmjs.com/package/@zhaow-de/puppeteer-cluster)
[![npm download count](https://img.shields.io/npm/dm/@zhaow-de/puppeteer-cluster)](https://www.npmjs.com/package/@zhaow-de/puppeteer-cluster)
[![Coverage Status](https://coveralls.io/repos/github/thomasdondorf/puppeteer-cluster/badge.svg?branch=master)](https://coveralls.io/github/thomasdondorf/puppeteer-cluster?branch=master)
[![Known Vulnerabilities](https://snyk.io/test/github/thomasdondorf/puppeteer-cluster/badge.svg)](https://snyk.io/test/github/thomasdondorf/puppeteer-cluster)
[![MIT License](https://img.shields.io/npm/l/puppeteer-cluster.svg)](#license)

Create a cluster of puppeteer workers. This library spawns a pool of Chromium instances via [Puppeteer] and helps to keep track of jobs and errors. This is helpful if you want to crawl multiple pages or run tests in parallel. Puppeteer Cluster takes care of reusing Chromium and restarting the browser in case of errors.

- [Installation](#installation)
- [Usage](#usage)
- [Examples](#examples)
- [Concurrency implementations](#concurrency-implementations)
- [Typings for input/output (via TypeScript Generics)](#typings-for-inputoutput-via-typescript-generics)
- [Debugging](#debugging)
- [API](#api)
- [License](#license)

###### What does this library do?

* Handling of crawling errors
* Auto restarts the browser in case of a crash
* Can automatically retry if a job fails
* Different concurrency models to choose from (pages, contexts, browsers)
* Simple to use, small boilerplate
* Progress view and monitoring statistics (see below)

<p align="center">
  <img src="https://i.imgur.com/koGNkBN.gif" height="250">
</p>

## Installation

Install using your favorite package manager:

```sh
npm install --save puppeteer # in case you don't already have it installed 
npm install --save puppeteer-cluster
```

Alternatively, use `yarn`:
```sh
yarn add puppeteer puppeteer-cluster
```

## Usage

The following is a typical example of using puppeteer-cluster. A cluster is created with 2 concurrent workers. Then a task is defined which includes going to the URL and taking a screenshot. We then queue two jobs and wait for the cluster to finish.

```js
const { Cluster } = require('puppeteer-cluster');

(async () => {
  const cluster = await Cluster.launch({
    concurrency: Cluster.CONCURRENCY_CONTEXT,
    maxConcurrency: 2,
  });

  await cluster.task(async ({ page, data: url }) => {
    await page.goto(url);
    const screen = await page.screenshot();
    // Store screenshot, do something else
  });

  cluster.queue('http://www.google.com/');
  cluster.queue('http://www.wikipedia.org/');
  // many more pages

  await cluster.idle();
  await cluster.close();
})();
```

## Examples
* [Simple example](examples/minimal.js)
* [Wait for a task to be executed](examples/execute.js)
* [Minimal screenshot server with express](examples/express-screenshot.js)
* [Deep crawling the Google search results](examples/deep-google-crawler.js)
* [Crawling the Alexa Top 1 Million](examples/alexa-1m.js)
* [Queuing functions (simple)](examples/function-queuing-simple.js)
* [Queuing functions (complex)](examples/function-queuing-complex.js)
* [Error handling](examples/error-handling.js)
* [Using a different puppeteer library (like puppeteer-core or puppeteer-firefox)](examples/different-puppeteer-library.js)
* [Provide types for input/output with TypeScript generics](examples/typings.ts)

## Concurrency implementations

There are different concurrency models, which define how isolated each job is run. You can set it in the `options` when calling [Cluster.launch](#Clusterlaunchoptions). The default option is `Cluster.CONCURRENCY_CONTEXT`, but it is recommended to always specify which one you want to use.

| Concurrency | Description | Shared data |
| --- | --- | --- |
| `CONCURRENCY_PAGE` | One [Page] for each URL | Shares everything (cookies, localStorage, etc.) between jobs. |
| `CONCURRENCY_CONTEXT` | Incognito page (see [IncognitoBrowserContext](https://github.com/GoogleChrome/puppeteer/blob/v1.5.0/docs/api.md#browsercreateincognitobrowsercontext)) for each URL  | No shared data. |
| `CONCURRENCY_BROWSER` | One browser (using an incognito page) per URL. If one browser instance crashes for any reason, this will not affect other jobs. | No shared data.  |
| Custom concurrency (**experimental**) | You can create your own concurrency implementation. Copy one of the files of the `concurrency/built-in` directory and implement `ConcurrencyImplementation`. Then provide the class to the option `concurrency`. **This part of the library is currently experimental and might break in the future, even in a minor version upgrade while the version has not reached 1.0.** | Depends on your implementation |

## Typings for input/output (via TypeScript Generics)

To allow proper type checks with TypeScript you can provide generics. In case no types are provided, `any` is assumed for input and output. See the following minimal example or check out the more complex [typings example](examples/typings.ts) for more information.

```ts
  const cluster: Cluster<string, number> = await Cluster.launch(/* ... */);

  await cluster.task(async ({ page, data }) => {
    // TypeScript knows that data is a string and expects this function to return a number
    return 123;
  });

  // Typescript expects a string as argument ...
  cluster.queue('http://...');

  // ... and will return a number when execute is called.
  const result = await cluster.execute('https://www.google.com');
```


## Debugging

Try to checkout the [puppeteer debugging tips](https://github.com/GoogleChrome/puppeteer#debugging-tips) first. Your problem might not be related to `puppeteer-cluster`, but `puppteer` itself. Additionally, you can enable verbose logging to see which data is consumed by which worker and some other cluster information. Set the DEBUG environment variable to `puppeteer-cluster:*`. See an example below or checkout the [debug docs](https://github.com/visionmedia/debug#windows-command-prompt-notes) for more information.

```bash
# Linux
DEBUG='puppeteer-cluster:*' node examples/minimal
# Windows Powershell
$env:DEBUG='puppeteer-cluster:*';node examples/minimal
```

## API

- [class: Cluster](#class-cluster)
  * [Cluster.launch(options)](#clusterlaunchoptions)
  * [cluster.task(taskFunction)](#clustertasktaskfunction)
  * [cluster.queue([data] [, taskFunction])](#clusterqueuedata--taskfunction)
  * [cluster.execute([data] [, taskFunction])](#clusterexecutedata--taskfunction)
  * [cluster.idle()](#clusteridle)
  * [cluster.close()](#clusterclose)

### class: Cluster

Cluster module provides a method to launch a cluster of Chromium instances.

#### event: 'taskerror'
- <[Error]>
- <[string]|[Object]>
- <[boolean]>

Emitted when a queued task ends in an error for some reason. Reasons might be a network error, your code throwing an error, timeout hit, etc. The first argument will the error itself. The second argument is the URL or data of the job (as given to [Cluster.queue]). If retryLimit is set to a value greater than `0`, the cluster will automatically requeue the job and retry it again later. The third argument is a boolean which indicates whether this task will be retried.
In case the task was queued via [Cluster.execute] there will be no event fired.

```js
  cluster.on('taskerror', (err, data, willRetry) => {
      if (willRetry) {
        console.warn(`Encountered an error while crawling ${data}. ${err.message}\nThis job will be retried`);
      } else {
        console.error(`Failed to crawl ${data}: ${err.message}`);
      }
  });
```

#### event: 'queue'
- <\?[Object]>
- <\?[function]>

Emitted when a task is queued via [Cluster.queue] or [Cluster.execute]. The first argument is the object containing the data (if any data is provided). The second argument is the queued function (if any). In case only a function is provided via [Cluster.queue] or [Cluster.execute], the first argument will be undefined. If only data is provided, the second argument will be undefined.

#### Cluster.launch(options)
- `options` <[Object]> Set of configurable options for the cluster. Can have the following fields:
  - `concurrency` <*Cluster.CONCURRENCY_PAGE*|*Cluster.CONCURRENCY_CONTEXT*|*Cluster.CONCURRENCY_BROWSER*|ConcurrencyImplementation> The chosen concurrency model. See [Concurreny models](#concurreny-models) for more information. Defaults to `Cluster.CONCURRENCY_CONTEXT`. Alternatively you can provide a class implementing `ConcurrencyImplementation`.
  - `maxConcurrency` <[number]> Maximal number of parallel workers. Defaults to `1`.
  - `puppeteerOptions` <[Object]> Object passed to [puppeteer.launch]. See puppeteer documentation for more information. Defaults to `{}`.
  - `perBrowserOptions` <[Array]<[Object]>> Object passed to [puppeteer.launch] for each individual browser. If set, `puppeteerOptions` will be ignored. Defaults to `undefined` (meaning that `puppeteerOptions` will be used).
  - `retryLimit` <[number]> How often do you want to retry a job before marking it as failed. Ignored by tasks queued via [Cluster.execute]. Defaults to `0`.
  - `retryDelay` <[number]> How much time should pass at minimum between the job execution and its retry. Ignored by tasks queued via [Cluster.execute]. Defaults to `0`.
  - `sameDomainDelay` <[number]> How much time should pass at minimum between two requests to the same domain. If you use this field, the queued `data` must be your URL or `data` must be an object containing a field called `url`.
  - `skipDuplicateUrls` <[boolean]> If set to `true`, will skip URLs which were already crawled by the cluster. Defaults to `false`. If you use this field, the queued `data` must be your URL or `data` must be an object containing a field called `url`.
  - `timeout` <[number]> Specify a timeout for all tasks. Defaults to `30000` (30 seconds).
  - `monitor` <[boolean]> If set to `true`, will provide a small command line output to provide information about the crawling process. Defaults to `false`.
  - `workerCreationDelay` <[number]> Time between creation of two workers. Set this to a value like `100` (0.1 seconds) in case you want some time to pass before another worker is created. You can use this to prevent a network peak right at the start. Defaults to `0` (no delay).
  - `puppeteer` <[Object]> In case you want to use a different puppeteer library (like [puppeteer-core](https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#puppeteer-vs-puppeteer-core) or [puppeteer-extra](https://github.com/berstend/puppeteer-extra)), pass the object here. If not set, will default to using puppeteer. When using `puppeteer-core`, make sure to also provide `puppeteerOptions.executablePath`.
- returns: <[Promise]<[Cluster]>>

The method launches a cluster instance.

#### cluster.task(taskFunction)
- `taskFunction` <[function]([string]|[Object], [Page], [Object])> Sets the function, which will be called for each job. The function will be called with an object having the following fields:
  - `page` <[Page]> The page given by puppeteer, which provides methods to interact with a single tab in Chromium.
  - `data` <any> The data of the job you provided to [Cluster.queue].
  - `worker` <[Object]> An object containing information about the worker executing the current job.
    - `id` <[number]> ID of the worker. Worker IDs start at 0.
- returns: <[Promise]>

Specifies a task for the cluster. A task is called for each job you queue via [Cluster.queue]. Alternatively you can directly queue the function that you want to be executed. See [Cluster.queue] for an example.

#### cluster.queue([data] [, taskFunction])
- `data` <any> Data to be queued. This might be your URL (a string) or a more complex object containing data. The data given will be provided to your task function(s). See [examples] for a more complex usage of this argument.
- `taskFunction` <[function]> Function like the one given to [Cluster.task]. If a function is provided, this function will be called (only for this job) instead of the function provided to [Cluster.task]. The function will be called with an object having the following fields:
  - `page` <[Page]> The page given by puppeteer, which provides methods to interact with a single tab in Chromium.
  - `data` <any> The data of the job you provided as first argument to [Cluster.queue]. This might be `undefined` in case you only specified a function.
  - `worker` <[Object]> An object containing information about the worker executing the current job.
    - `id` <[number]> ID of the worker. Worker IDs start at 0.
- returns: <[Promise]>

Puts a URL or data into the queue. Alternatively (or even additionally) you can queue functions. See the examples about function queuing for more information: ([Simple function queuing](examples/function-queuing-simple.js), [complex function queuing](examples/function-queuing-complex.js)).

Be aware that this function only returns a Promise for backward compatibility reasons. This function does not run asynchronously and will immediately return.

#### cluster.execute([data] [, taskFunction])
- `data` <any> Data to be queued. This might be your URL (a string) or a more complex object containing data. The data given will be provided to your task function(s). See [examples] for a more complex usage of this argument.
- `taskFunction` <[function]> Function like the one given to [Cluster.task]. If a function is provided, this function will be called (only for this job) instead of the function provided to [Cluster.task]. The function will be called with an object having the following fields:
  - `page` <[Page]> The page given by puppeteer, which provides methods to interact with a single tab in Chromium.
  - `data` <any> The data of the job you provided as first argument to [Cluster.queue]. This might be `undefined` in case you only specified a function.
  - `worker` <[Object]> An object containing information about the worker executing the current job.
    - `id` <[number]> ID of the worker. Worker IDs start at 0.
- returns: <[Promise]>

Works like [Cluster.queue], but this function returns a Promise which will be resolved after the task is executed. That means, that the job is still queued, but the script will wait for it to be finished. In case an error happens during the execution, this function will reject the Promise with the thrown error. There will be no "taskerror" event fired. In addition, tasks queued via execute will ignore "retryLimit" and "retryDelay". For an example see the [Execute example](examples/execute.js).

#### cluster.idle()
- returns: <[Promise]>

Promise is resolved when the queue becomes empty.

#### cluster.close()
- returns: <[Promise]>

Closes the cluster and all opened Chromium instances including all open pages (if any were opened). It is recommended to run [Cluster.idle](#clusteridle) before calling this function. The [Cluster] object itself is considered to be disposed and cannot be used anymore.

## License

[MIT license](./LICENSE).



[Cluster.queue]: #clusterqueuedata--taskfunction "Cluster.queue"
[Cluster.execute]: #clusterexecutedata--taskfunction "Cluster.execute"
[Cluster.task]: #clustertasktaskfunction "Cluster.task"
[Cluster]: #class-cluster "Cluster"

[Puppeteer]: https://github.com/GoogleChrome/puppeteer "Puppeteer"
[Page]: https://github.com/GoogleChrome/puppeteer/blob/v1.5.0/docs/api.md#class-page "Page"
[puppeteer.launch]: https://github.com/GoogleChrome/puppeteer/blob/v1.5.0/docs/api.md#puppeteerlaunchoptions "puppeteer.launch"

[function]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function "Function"
[string]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type "String"
[number]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Number_type "Number"
[Promise]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise "Promise"
[boolean]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Boolean_type "Boolean"
[Object]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object "Object"
[Error]: https://nodejs.org/api/errors.html#errors_class_error "Error"
[Array]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array "Array"


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [0.30.0, upstream-0.23.0] - 2022-01-23
### Changed
- Updated dependencies which haven't been updated for more than one year
- Moved from Travis to Github Actions
- Remove the linter for now, but will probably add it back later after having a look at the pull requests.
### Fixed
- Fixed bug #395 not using `perBrowserOptions`

## [0.22.0] - 2020-08-06
### Changed
- Updated dependencies to their latest versions
  - In particular the devDependencies  `puppeteer` and `puppeteer-core` received a major upgrade (to version `5`)
  - `peerDependencies` was changed accordingly.

## [0.21.0] - 2020-05-24
### Changed
- Updated dependencies to their latest versions
  - In particular the devDependencies  `puppeteer` and `puppeteer-core` received a major upgrade (to version `3`)
- `peerDependencies` was changed to `^1.5.0 || ^2.0.0 || ^3.0.0` as there were no changes regarding the API that this library uses.
- Removed support for Node.js version 8 as puppeteer [dropped support](https://github.com/puppeteer/puppeteer/releases/tag/v3.0.0) for it.
- Added check to ensure `maxConcurrency` is set (#243)

## [0.20.0] - 2020-03-26
### Added
- Add a new option `perBrowserOptions` for using different args for each puppeteer instance
  - Allowing to use a different proxy for each worker
### Changed
- Updated dev dependencies to their latest versions

## [0.19.0] - 2020-02-16
### Changed
- Added the parameter `jobWillRetry` to the `'taskerror'` Event

## [0.18.0] - 2019-12-05
### Changed
- Updated dependencies to their latest versions
  - In particular the devDependencies  `puppeteer` and `puppeteer-core` received a major upgrade (to version `2`)
- `peerDependencies` was changed to `^1.5.0 || ^2.0.0` as there were no changes regarding the API that this library uses.
- Removed support for Node.js version 6 as puppeteer [dropped support](https://github.com/puppeteer/puppeteer/pull/5045) for it.

## [0.17.0] - 2019-08-03
### Changed
- Updated dependencies to their latest versions
### Fixed
- Fixed bug making it impossible to queue `null` (issue #178)

## [0.16.0] - 2019-05-11
### Changed
- License changed to MIT
- Updated dependencies to their latest versions

## [0.15.2] - 2019-03-09
### Fixed
- Fixed bug, which crashed the cluster in some environments (#113)

## [0.15.1] - 2019-03-06
### Changed
- Changed when the `queue` event is emitted (after the job is queued into the internal queue).

## [0.15.0] - 2019-03-06
### Added
- Event `queue` added

## [0.14.0] - 2019-02-28
### Added
- Support for generics via `Cluster<InType, OutType>`

## [0.13.2] - 2019-02-26
### Fixed
- `package-lock.json` file was not updated

## [0.13.1] - 2019-02-26
### Fixed
- Internal helper function was accidentally set to public

## [0.13.0] - 2019-02-25
### Added
- `Cluster.execute` function added
- Examples for `Cluster.execute` added

## [0.12.1] - 2018-11-08
### Fixed
- Fixed support for custom concurrency implementations

## [0.12.0] - 2018-11-07
### Added
- Support for custom puppeteer libraries added
- Support for custom concurrency implementations added
### Changed
- Updated dependencies to their latest versions

## [0.11.2] - 2018-09-07
### Fixed
- Fixed another sameDomainDelay bug leading to high CPU usage

## [0.11.1] - 2018-09-06
### Fixed
- Fixed `sameDomainDelay` bug (issue #11)

## [0.11.0] - 2018-09-05
### Fixed
- Fixed rarely happening bug (issue #3), which made browser not able to restart

## [0.10.0] - 2018-08-21
### Added
- Node.js support for version 6 and 7 added

## [0.9.1] - 2018-07-18
### Fixed
- Circular structures led to crashs in case of crawling errors.

## [0.9.0] - 2018-07-13
### Changed
- Cluster.task function signature changed from `Cluster.task(page, url)` to `Cluster.task({ page, data })`. `Cluster.queue` can be passed any data instead of a string or object.

## 0.8.1 - 2018-07-08
### Fixed
- The timeout-promise for a task was not canceled when a task threw an error.

## 0.8.0 - 2018-07-04
### Added
- Cluster can be used without defining a Cluster.task function by queuing only functions.

### Fixed
- Page errors ("Page crashed!") were not caught so far

### Removed
- maxCPU and maxMemory options removed as they made no sense (better to check how much chromium your machine can handle.)


test/util.test.ts

import { formatDateTime, formatDuration, timeoutExecute } from '../src/util';

describe('formatDateTime', () => {

    test('formats dates', () => {
        expect(
            formatDateTime(new Date(1980, 11, 31, 18, 30, 51, 999)),
        ).toBe('1980-12-31 18:30:51.999');
    });

    test('zero-fills dates', () => {
        expect(
            formatDateTime(new Date(2018, 0, 7, 4, 5, 7, 12)),
        ).toBe('2018-01-07 04:05:07.012');
    });

    test('works also for numbers', () => {
        const now = new Date();
        const nowTimestamp = now.getTime();
        const nowStr = formatDateTime(now);
        expect(
            formatDateTime(nowTimestamp),
        ).toBe(nowStr);
    });

});

describe('formatDuration', () => {

    test('negative', () => {
        expect(formatDuration(-123))
        .toBe('unknown');
    });
    test('round up', () => {
        expect(formatDuration(995))
        .toBe('1.0 seconds');
    });
    test('milliseconds 1', () => {
        expect(formatDuration(0))
        .toBe('0.0 ms');
    });
    test('milliseconds 2', () => {
        expect(formatDuration(800))
        .toBe('800.0 ms');
    });
    test('seconds', () => {
        expect(formatDuration(10.29 * 1000))
        .toBe('10.3 seconds');
    });
    test('minutes', () => {
        expect(formatDuration(10.29 * 60 * 1000))
        .toBe('10.3 minutes');
    });
    test('hours', () => {
        expect(formatDuration(10.29 * 60 * 60 * 1000))
        .toBe('10.3 hours');
    });
    test('days', () => {
        expect(formatDuration(10.29 * 24 * 60 * 60 * 1000))
        .toBe('10.3 days');
    });
    test('months', () => {
        expect(formatDuration(10.29 * 31 * 24 * 60 * 60 * 1000))
        .toBe('10.3 months');
    });
    test('years', () => {
        expect(formatDuration(10.29 * 365 * 24 * 60 * 60 * 1000))
        .toBe('10.3 years');
    });

});

describe('timeoutExecute', () => {
    jest.useRealTimers();

    test('timeout after given time', async () => {
        expect.assertions(1);

        try {
            await timeoutExecute(
                10,
                new Promise(resolve => setTimeout(resolve, 50)),
            );
            fail();
        } catch (err: any) {
            expect(err.message).toEqual(expect.stringContaining('Timeout'));
        }
    });

    test('execute promise before timeout', async () => {
        const result = await timeoutExecute(
            50,
            new Promise(resolve => setTimeout(() => resolve(123), 10)),
        );
        expect(result).toBe(123);
    });

    test('cancel timeout when promise resolves', async () => {
        const result = await timeoutExecute(
            100000,
            new Promise(resolve => setTimeout(() => resolve(123), 10)),
        );
        expect(result).toBe(123);
    });

});


src/Worker.ts
import Job from './Job';
import type Cluster from './Cluster';
import type { TaskFunction } from './Cluster';
import type { Page } from 'puppeteer';
import { timeoutExecute, debugGenerator, log } from './util';
import { inspect } from 'util';
import { WorkerInstance, JobInstance } from './concurrency/ConcurrencyImplementation';

const debug = debugGenerator('Worker');

interface WorkerOptions {
    cluster: Cluster;
    args: string[];
    id: number;
    browser: WorkerInstance;
}

const BROWSER_INSTANCE_TRIES = 10;

export interface WorkError {
    type: 'error';
    error: Error;
}

export interface WorkData {
    type: 'success';
    data: any;
}

export type WorkResult = WorkError | WorkData;

export default class Worker<JobData, ReturnData> implements WorkerOptions {

    cluster: Cluster;
    args: string[];
    id: number;
    browser: WorkerInstance;

    activeTarget: Job<JobData, ReturnData> | null = null;

    public constructor({ cluster, args, id, browser }: WorkerOptions) {
        this.cluster = cluster;
        this.args = args;
        this.id = id;
        this.browser = browser;

        debug(`Starting #${this.id}`);
    }

    public async handle(
        task: TaskFunction<JobData, ReturnData>,
        job: Job<JobData, ReturnData>,
        timeout: number,
    ): Promise<WorkResult> {
        this.activeTarget = job;

        let jobInstance: JobInstance | null = null;
        let page: Page | null = null;

        let tries = 0;

        while (jobInstance === null) {
            try {
                jobInstance = await this.browser.jobInstance();
                page = jobInstance.resources.page;
            } catch (err: any) {
                debug(`Error getting browser page (try: ${tries}), message: ${err.message}`);
                await this.browser.repair();
                tries += 1;
                if (tries >= BROWSER_INSTANCE_TRIES) {
                    throw new Error('Unable to get browser page');
                }
            }
        }

        // We can be sure that page is set now, otherwise an exception would've been thrown
        page = page as Page; // this is just for TypeScript

        let errorState: Error | null = null;

        page.on('error', (err) => {
            errorState = err;
            log(`Error (page error) crawling ${inspect(job.data)} // message: ${err.message}`);
        });

        debug(`Executing task on worker #${this.id} with data: ${inspect(job.data)}`);

        let result: any;
        try {
            result = await timeoutExecute(
                timeout,
                task({
                    page,
                    // data might be undefined if queue is only called with a function
                    // we ignore that case, as the user should use Cluster<undefined> in that case
                    // to get correct typings
                    data: job.data as JobData,
                    worker: {
                        id: this.id,
                    },
                }),
            );
        } catch (err: any) {
            errorState = err;
            log(`Error crawling ${inspect(job.data)} // message: ${err.message}`);
        }

        debug(`Finished executing task on worker #${this.id}`);

        try {
            await jobInstance.close();
        } catch (e: any) {
            debug(`Error closing browser instance for ${inspect(job.data)}: ${e.message}`);
            await this.browser.repair();
        }

        this.activeTarget = null;

        if (errorState) {
            return {
                type: 'error',
                error: errorState || new Error('asf'),
            };
        }
        return {
            data: result,
            type: 'success',
        };
    }

    public async close(): Promise<void> {
        try {
            await this.browser.close();
        } catch (err: any) {
            debug(`Unable to close worker browser. Error message: ${err.message}`);
        }
        debug(`Closed #${this.id}`);
    }

}


src/SystemMonitor.ts
import * as os from 'os';

interface SystemLoad {
    idle: number;
    total: number;
}

const INIT_INTERVAL = 50;
const MEASURE_INTERVAL = 200;

// timespan of which to measure load
// must be a multiple of MEASURE_INTERVAL
const MEASURE_TIMESPAN = 5000;

const loadListSize = MEASURE_TIMESPAN / MEASURE_INTERVAL;

export default class SystemMonitor {

    private cpuUsage: number = 0;
    private memoryUsage: number = 0;

    private loads: SystemLoad[] = [];

    private interval: NodeJS.Timer | null = null;

    // After init is called there is at least something in the cpuUsage thingy
    public init() {
        this.calcLoad();
        return new Promise<void>((resolve) => {
            setTimeout(
                () => {
                    this.calcLoad();
                    this.interval = setInterval(() => this.calcLoad(), MEASURE_INTERVAL);
                    resolve();
                },
                INIT_INTERVAL,
            );
        });
    }

    public close() {
        clearInterval(this.interval as NodeJS.Timer);
    }

    private calcLoad() { // based on https://gist.github.com/bag-man/5570809
        let totalIdle = 0;
        let totalTick = 0;
        const cpus = os.cpus();

        /* istanbul ignore next */
        if (!cpus) {
            // In some environments, os.cpus() might return undefined (although it's not stated in
            // the Node.js docs), see #113 for more information
            return;
        }

        for (let i = 0, len = cpus.length; i < len; i += 1) {
            const cpu = cpus[i];
            for (const type in cpu.times) {
                totalTick += (cpu.times as any)[type];
            }
            totalIdle += cpu.times.idle;
        }

        const currentLoad = {
            idle: totalIdle / cpus.length,
            total: totalTick / cpus.length,
        };

        if (this.loads.length !== 0) {
            const compareLoad = this.loads[0];
            const idleDifference = currentLoad.idle - compareLoad.idle;
            const totalDifference = currentLoad.total - compareLoad.total;

            this.cpuUsage = 100 - (100 * idleDifference / totalDifference);
            this.memoryUsage = 100 - (100 * os.freemem() / os.totalmem());
        }

        this.loads.push(currentLoad);
        if (this.loads.length > loadListSize) {
            // remove oldest entry
            this.loads.shift();
        }
    }

    public getCpuUsage() {
        return this.cpuUsage;
    }
    public getMemoryUsage() {
        return this.memoryUsage;
    }
}


src/concurrency/ConcurrencyImplementation.ts

import { Page, LaunchOptions, BrowserLaunchArgumentOptions, BrowserConnectOptions } from 'puppeteer';

/**
 * ABSTRACT CLASS Needs to be implemented to manage one or more browsers via puppeteer instances
 *
 * The ConcurrencyImplementation creates WorkerInstances. Workers create JobInstances:
 * One WorkerInstance per maxWorkers, one JobInstance per job
 */
export default abstract class ConcurrencyImplementation {

    protected options: LaunchOptions;
    protected puppeteer: any;

    /**
     * @param options  Options that should be provided to puppeteer.launch
     * @param puppeteer  puppeteer object (like puppeteer or puppeteer-core)
     */
    public constructor(options: LaunchOptions, puppeteer: any) {
        this.options = options;
        this.puppeteer = puppeteer;
    }

    /**
     * Initializes the manager
     */
    public abstract init(): Promise<void>;

    /**
     * Closes the manager (called when cluster is about to shut down)
     */
    public abstract close(): Promise<void>;

    /**
     * Creates a worker and returns it
     */
    public abstract workerInstance(perBrowserOptions: LaunchOptions & BrowserLaunchArgumentOptions & BrowserConnectOptions | undefined):
        Promise<WorkerInstance>;

}

/**
 * WorkerInstances are created by calling the workerInstance function.
 * In case maxWorkers is set to 4, 4 workers will be created.
 */
export interface WorkerInstance {
    jobInstance: () => Promise<JobInstance>;

    /**
     * Closes the worker (called when the cluster is about to shut down)
     */
    close: () => Promise<void>;

    /**
     * Repair is called when there is a problem with the worker (like call or close throwing
     * an error)
     */
    repair: () => Promise<void>;
}

/**
 * JobInstance which is created for the execution of one job. After usage
 * the associated resources will be destroyed by calling close.
 * resources needs to contain the page and might contain any other
 * related resources (like the browser).
 */
export interface JobInstance {
    resources: ResourceData;

    /**
     * Called to close the related resources
     */
    close: () => Promise<void>;
}

export interface ResourceData {
    page: Page;
    [key: string]: any;
}

export type ConcurrencyImplementationClassType = new (
    options: LaunchOptions,
    puppeteer: any,
) => ConcurrencyImplementation;


src/util.ts

import * as Debug from 'debug';

interface TimeUnit {
    step: number;
    name: string;
}

function timeUnit(step: number, name: string) : TimeUnit {
    return { step, name };
}

const TIME_UNITS: TimeUnit[] = [
    timeUnit(1, 'ms'),
    timeUnit(1000, 'seconds'),
    timeUnit(60, 'minutes'),
    timeUnit(60, 'hours'),
    timeUnit(24, 'days'),
    timeUnit(31, 'months'),
    timeUnit((365 / 31), 'years'),
];

const TIME_UNIT_THRESHOLD = 0.95;

function padDate(value: number|string, num: number): string {
    const str = value.toString();
    if (str.length >= num) {
        return str;
    }
    const zeroesToAdd = num - str.length;
    return '0'.repeat(zeroesToAdd) + str;
}

export function formatDateTime(datetime: Date | number): string {
    const date = (typeof datetime === 'number') ? new Date(datetime) : datetime;

    const dateStr = `${date.getFullYear()}`
        + `-${padDate(date.getMonth() + 1, 2)}`
        + `-${padDate(date.getDate(), 2)}`;
    const timeStr = `${padDate(date.getHours(), 2)}`
        + `:${padDate(date.getMinutes(), 2)}`
        + `:${padDate(date.getSeconds(), 2)}`
        + `.${padDate(date.getMilliseconds(), 3)}`;

    return `${dateStr} ${timeStr}`;
}

export function formatDuration(millis: number): string {
    if (millis < 0) {
        return 'unknown';
    }

    let remaining = millis;
    let nextUnitIndex = 1;
    while (nextUnitIndex < TIME_UNITS.length &&
    remaining / TIME_UNITS[nextUnitIndex].step >= TIME_UNIT_THRESHOLD) {
        remaining = remaining / TIME_UNITS[nextUnitIndex].step;
        nextUnitIndex += 1;
    }

    return `${remaining.toFixed(1)} ${TIME_UNITS[nextUnitIndex - 1].name}`;
}

export async function timeoutExecute<T>(millis: number, promise: Promise<T>): Promise<T> {

    let timeout: NodeJS.Timer | null = null;

    const result = await Promise.race([
        (async () => {
            await new Promise((resolve) => {
                timeout = setTimeout(resolve, millis);
            });
            throw new Error(`Timeout hit: ${millis}`);
        })(),
        (async () => {
            try {
                return await promise;
            } catch (error: any) {
                // Cancel timeout in error case
                clearTimeout(timeout as any as NodeJS.Timer);
                throw error;
            }
        })(),
    ]);
    clearTimeout(timeout as any as NodeJS.Timer); // is there a better way?
    return result;
}

export function debugGenerator(namespace: string): Debug.IDebugger {
    const debug = Debug(`puppeteer-cluster: ${namespace}`);
    return debug;
}

const logToConsole = Debug('puppeteer-cluster:log');
logToConsole.log = console.error.bind(console);

export function log(msg: string): void {
    logToConsole(msg);
}
